python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py

```diff
@@ -146,16 +159,198 @@ def get_last_updated_timestamp(record: Optional[EventLogRecord]) -> Optional[flo
+    freshness_cron_timezone: str,
+    severity: AssetCheckSeverity,
+    lower_bound_delta: datetime.timedelta,
+    asset_property_enforcement_lambda: Optional[Callable[[AssetsDefinition], bool]],
+) -> AssetChecksDefinition:
+    params_metadata = {}
+    if deadline_cron:
+        params_metadata[FRESHNESS_CRON_METADATA_KEY] = deadline_cron
+        params_metadata[FRESHNESS_CRON_TIMEZONE_METADATA_KEY] = freshness_cron_timezone
+    if lower_bound_delta:
+        params_metadata[LOWER_BOUND_DELTA_METADATA_KEY] = lower_bound_delta.total_seconds()
+
+    @asset_check(
+        asset=asset_key,
+        description="Evaluates freshness for targeted asset.",
+        name="freshness_check",
+        metadata={FRESHNESS_PARAMS_METADATA_KEY: params_metadata},
+    )
+    def the_check(context: AssetCheckExecutionContext) -> AssetCheckResult:
+        if asset_property_enforcement_lambda:
+            asset_property_enforcement_lambda(
+                context.job_def.asset_layer.asset_graph.get(asset_key).assets_def
+            )
+        current_timestamp = pendulum.now("UTC").timestamp()
+
+        # Explicit call to partitions def here will be replaced with AssetSlice reference once it's available.
+        partitions_def = cast(
+            Optional[TimeWindowPartitionsDefinition],
+            context.job_def.asset_layer.asset_graph.get(asset_key).partitions_def,
+        )
+        check.invariant(
+            partitions_def is None or isinstance(partitions_def, TimeWindowPartitionsDefinition),
+            "Expected partitions_def to be time-windowed.",
+        )
+
+        last_update_time_lower_bound = get_last_update_time_lower_bound(
+            deadline_cron=deadline_cron,
+            freshness_cron_timezone=freshness_cron_timezone,
+            current_timestamp=current_timestamp,
+            lower_bound_delta=lower_bound_delta,
+        )
+        expected_partition_key = get_latest_complete_partition_key(
+            deadline_cron=deadline_cron,
+            current_timestamp=current_timestamp,
+            freshness_cron_timezone=freshness_cron_timezone,
+            partitions_def=partitions_def,
+        )
+        latest_record = retrieve_latest_record(
+            instance=context.instance, asset_key=asset_key, partition_key=expected_partition_key
+        )
+        update_timestamp = get_last_updated_timestamp(latest_record)
+        passed = (
+            update_timestamp is not None
+            and update_timestamp >= last_update_time_lower_bound.timestamp()
+        )
+
+        return AssetCheckResult(
+            passed=passed,
+            description=get_description_for_freshness_check(
+                passed,
+                update_timestamp,
+                last_update_time_lower_bound,
+                current_timestamp,
+                expected_partition_key,
+            ),
+            severity=severity,
+        )
+
+    return the_check
+
+
+def get_description_for_freshness_check(
```

Comment:
```suggestion
def get_description_for_freshness_check_result(
```

Reply:
ah yea I think i actually changed to this in the upstream. but agreed.